traits/trait_dict_object.py

```diff
@@ -325,21 +325,26 @@ def setdefault(self, key, value=None):
                 Will be an empty dict.
 
         """
+        if key in self:
+            return self[key]
 
-        added = None
-        if key not in self:
-            key = self.validate_key(key)
-            value = self.validate_value(value)
+        key = self.validate_key(key)
+        value = self.validate_value(value)
+
+        if key in self:
```

Comment:
We first check if the raw key is in dict, if not, we transform it (by calling validate) and check for presence, but if the transformed key is present, we don't return its value, instead we call `__setitem__` again. I'm not sure if this is the intended behavior.

Reply:
Transformation of keys make the operations weird. Because we can't forbid the use of transformation, we need to deal with it in the most consistent way possible.
Consider `setdefault` being translated to `__getitem__` ---(if failed)---> `__setitem__` ---> `__getitem__`. We don't transform the keys for `__getitem__` operations anywhere else.